src/blocks/battery/upower.rs

```diff
@@ -57,86 +56,176 @@ impl Device {
+                    changes,
+                })
+            }
+            None => None,
+        })
+    }
+}
+
+pub(super) struct Device {
+    dbus_conn: Connection,
+    device: DeviceName,
+    device_conn: Option<DeviceConnection>,
+    device_added_stream: MessageStream,
+    device_removed_stream: MessageStream,
+}
+
+impl Device {
+    pub(super) async fn new(device: DeviceName) -> Result<Self> {
+        let dbus_conn = new_system_dbus_connection().await?;
+
+        let device_connection = DeviceConnection::new(&dbus_conn, &device).await?;
+
+        let device_added_stream = MessageStream::for_match_rule(
+            MatchRule::builder()
+                .msg_type(zbus::MessageType::Signal)
+                .interface("org.freedesktop.UPower")
+                .and_then(|x| x.path("/org/freedesktop/UPower"))
+                .and_then(|x| x.member("DeviceAdded"))
+                .unwrap()
+                .build(),
+            &dbus_conn,
+            None,
+        )
+        .await
+        .error("Failed to add match rule")?;
+
+        let device_removed_stream = MessageStream::for_match_rule(
+            MatchRule::builder()
+                .msg_type(zbus::MessageType::Signal)
+                .interface("org.freedesktop.UPower")
+                .and_then(|x| x.path("/org/freedesktop/UPower"))
+                .and_then(|x| x.member("DeviceRemoved"))
+                .unwrap()
+                .build(),
+            &dbus_conn,
+            None,
+        )
+        .await
+        .error("Failed to add match rule")?;
 
         Ok(Self {
-            device_proxy,
-            changes,
+            dbus_conn,
+            device,
+            device_conn: device_connection,
+            device_added_stream,
+            device_removed_stream,
         })
     }
 }
 
 #[async_trait]
 impl BatteryDevice for Device {
     async fn get_info(&mut self) -> Result<Option<BatteryInfo>> {
-        let capacity = self
-            .device_proxy
-            .percentage()
-            .await
-            .error("Failed to get capacity")?;
-
-        let power = self
-            .device_proxy
-            .energy_rate()
-            .await
-            .error("Failed to get power")?;
-
-        let status = match self
-            .device_proxy
-            .state()
-            .await
-            .error("Failed to get status")?
-        {
-            1 => BatteryStatus::Charging,
-            2 | 6 => BatteryStatus::Discharging,
-            3 => BatteryStatus::Empty,
-            4 => BatteryStatus::Full,
-            5 => BatteryStatus::NotCharging,
-            _ => BatteryStatus::Unknown,
-        };
+        match &self.device_conn {
+            None => Ok(None),
+            Some(device_connection) => {
+                let capacity = device_connection
+                    .device_proxy
+                    .percentage()
+                    .await
+                    .error("Failed to get capacity")?;
```

Comment:
If this error happens it probably means that device suddenly became unavailable. This can be either just assumed (bluetooth block [does this](https://github.com/greshake/i3status-rust/blob/c3fc0f0d50871089c6718d335c1234533cd4ecbd/src/blocks/bluetooth.rs#L259-L267)) or the error variant could be checked (I haven't checked which error variant is returned in this case).